lib/recursive-methods.rb

```diff
@@ -1,49 +1,86 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) run n times to get all the nums
+# Space complexity: O(n) call stack memory
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    return 1 if n <= 1
+    return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n^2): O(n) string concatenation * O(n) recursive operation
+# Space complexity: O(n^2): O(n) call stack memory * O(n) new string
+def reverse(s, index = s.length - 1)
+    return s[0] || "" if index <= 0
+    return s[index] + reverse(s, index - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n): n being the length of s
+# Space complexity: O(n) call stack memory
+def reverse_inplace(s, i = 0, j = s.length - 1)
+    return s || ""  if i >= j
+    swap(s, i, j)
+    return reverse_inplace(s, i + 1, j - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def swap(s, i, j)
+    temp = s[i]
+    s[i] = s[j]
+    s[j] = temp
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n) call stack memory
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return bunny(n - 1) + 2
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n) call stack memory
+def nested(s, i = 0, j = s.length - 1)
+    return true if i == j - 1 && s[i] != s[j] || s == ""
+    return false if s[i] == s[j] || s.length.odd?
+    return nested(s, i + 1, j - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n) call stack memory
+def search(array, value, index = 0)
+    return true if array[index] == value
+    return false if index >= array.length
+    return search(array, value, index + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n) call stack memory
+def is_palindrome(s, i = 0, j = s.length - 1)
+    return true if i >= j && s[i] == s[j] || s == ""
+    return false if s[i] != s[j]
+    return is_palindrome(s, i + 1, j - 1)
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+def digit_match(n, m, index = 0, matches = 0)
```

Comment:
👍 This works, but has a space complexity of O(n^2) and a similar time complexity.